components/base-rest-api/packages/services/lib/token-swapper-service.js

```diff
@@ -0,0 +1,117 @@
+import jwtDecode from 'jwt-decode';
+import { newInvoker } from './authentication-providers/helpers/invoker';
+
+const settingKeys = {
+  dbValidTokens: 'dbValidTokens',
+};
+class TokenSwapperService extends Service {
+  constructor() {
+    super();
+    this.boom.extend(['invalidToken', 403]);
+    this.dependency(['dbService', 'authenticationProviderConfigService']);
+  }
+
+  async init() {
+    this.invoke = newInvoker(this.container.find.bind(this.container));
+  }
+
+  async swap({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    if (!uid) {
+      throw this.boom.badRequest('Empty uid', true);
+    }
+    let payload;
+    try {
+      payload = jwtDecode(token);
+    } catch (error) {
+      throw this.boom.invalidToken('Invalid Token', true).cause(error);
+    }
+    const ttl = _.get(payload, 'exp', 0);
+    if (ttl === 0) {
+      throw this.boom.invalidToken('Expired Token', true);
+    }
+    const { isSameToken } = await this.revokeValidToken({ token, uid });
+    if (isSameToken) {
+      return;
+    }
+    const record = { uid, token, ttl };
+    await dbService.helper
+      .updater()
+      .table(dbValidTokens)
+      .key({ uid })
+      .item(record)
+      .update();
+  }
+
+  async revokeValidToken({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const authenticationProviderConfigService = await this.service('authenticationProviderConfigService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    const record = await dbService.helper
+      .getter()
+      .table(dbValidTokens)
+      .key({ uid })
+      .get();
+
+    if (!record) {
+      return { isSameToken: false };
+    }
+
+    const oldToken = record.token;
+    const isSameToken = token === oldToken;
+    if (isSameToken) {
+      return { isSameToken: true };
+    }
+
+    let payload;
+    try {
+      payload = jwtDecode(oldToken);
+    } catch (error) {
+      throw this.boom.invalidToken('Invalid Revoke Token', true).cause(error);
+    }
+    const providerId = payload.iss;
+    const providerConfig = await authenticationProviderConfigService.getAuthenticationProviderConfig(providerId);
+
+    const tokenRevokerLocator = _.get(providerConfig, 'config.type.config.impl.tokenRevokerLocator');
+    if (!tokenRevokerLocator) {
+      throw this.boom.badRequest(
+        `Error revoking token. The authentication provider with id = '${providerId}' does not support token revocation`,
+        false,
+      );
+    }
+
+    this.log.info('Revoking previous user session');
+    // invoke the token revoker and pass the token that needs to be revoked
+
+    try {
+      await this.invoke(tokenRevokerLocator, undefined, { token: oldToken }, providerConfig);
+    } catch (error) {
+      throw this.boom.badRequest(`Error trying to revoke token`, true).cause(error);
```

Comment:
- not entirely familiar with boom.cause, but since it sets the root cause of the error, I'm thinking it may be better to set safe to 'false' here, especially since we are in an authentication handler. If your testing has shown this is indeed safe to return to the user like this, I will yield to that (but I would also double check the return value in a debugger or test to be sure).
- There should be a unit test for a failed token revoke: simulate a bad request, network error, etc.